src/org/opensolaris/opengrok/util/OptionParser.java

```diff
@@ -0,0 +1,701 @@
+    
+    // Option object referenced by its name(s)
+    final private HashMap<String,Option> options;
+    
+    // List of options in order of declaration
+    final private List<Option> optionList;
+    
+    // Keeps track of separator elements placed in option summary
+    final private List<Object> usageSummary;
+    
+    private boolean scanning = false;
+    
+    public String prologue;  // text emitted before option summary
+    public String epilogue;  // text emitted after options summary
+    
+    class DuplicateOptionName extends Exception {
+        public DuplicateOptionName(String message) {
+            super(message);
+        }
+    }
+    
+    public class Option {
+        
+        List<String> names;          // option names/aliases
+        String argument;             // argument name for summary
+        String value;                // user entered value for option
+        Class<?> valueType;          // eg. Integer.class, other than String
+        Pattern valuePattern;        // pattern used to accept value
+        List<String> allowedValues;  // list of restricted values
+        Boolean mandatory;           // true/false when argument present
+        StringBuilder description;   // option description for summary
+        Consumer<Object> action;     // code to execute when option encountered
+        
+        public Option() {
+            names = new ArrayList<String>();
+        }
+                
+        public void set(String option, String arg) throws IllegalArgumentException {
+            addAlias(option);
+            setArgument(arg);
+        }
+        
+        public void addAlias(String alias) throws IllegalArgumentException {
+            names.add(alias);
+            
+            if (options.containsKey(alias)) {
+                Option aka = options.get(alias);
+                throw new IllegalArgumentException("** Programmer error! Option " + alias + " already defined");
+            }
+            
+            options.put(alias, this);
+        }
+        
+        public void setAllowedValues(String[] allowed) {
+            allowedValues = Arrays.asList(allowed);
+        }
+        
+        public void setValueType(Class<?> type){
+            valueType = type;
+        }
+        
+        public void setArgument(String arg) {
+            argument = arg.trim();
+            mandatory = !argument.startsWith("[");
+        }
+        
+        public void setPattern(String pattern) {
+            valuePattern = Pattern.compile(pattern);
+        }
+        
+        public void addDescription(String descrip) {
+            if (description == null) {
+                description = new StringBuilder();
+            }
+            description.append(descrip);
+            description.append("\n");
+        }
+        
+        /**
+         * Store user defined action with option object
+         * @param action 
+         */
+        public void Do(Consumer<Object> action) {
```

Comment:
`setAction` ?

Reply:
This one is seen/used by the caller and I want it to appear to be similar in name to the same named method in OptionParser. For example,

OptionParser openGrok = OptionParser.Do(parser -> {

    parser.on("-?", "-h", "--help", "Display this usage.").Do( v -> {
	parser.help();
    });

So when one reads the code they can think, 'on' seeing this option string, the code will 'do' the following...

With your suggestion, the code would look like this, which doesn't read the same.

OptionParser openGrok = OptionParser.Do(parser -> {

    parser.on("-?", "-h", "--help", "Display this usage.").setAction( v -> {
	parser.help();
    });

Reply:
I prefer `do` but I don't get why it's with a capital letter

Reply:
I would have chosen 'do', but the compiler didn't like that, it is a keyword to the java language.

Reply:
ah :-D